test/test_images/sender/main.go

```diff
@@ -81,3 +121,8 @@ func writeTerminationMessage(result interface{}) error {
 	}
 	return ioutil.WriteFile("/dev/termination-log", b, 0644)
 }
+
+// isRetryable determines if the err is an error which is retryable
+func isRetryable(err error) bool {
+	return strings.Contains(err.Error(), "404 Not Found") || strings.Contains(err.Error(), "503 Service Unavailable") || strings.Contains(err.Error(), "500 Internal Server Error")
```

Comment:
I would avoid retrying 500 errors.

Reply:
500 errors was caught in https://pantheon.corp.google.com/storage/browser/_details/knative-prow/pr-logs/pull/google_knative-gcp/1415/pull-google-knative-gcp-wi-tests/1281267387093487616/artifacts/pod-logs/test-g-c-p-broker-tracing-kqp86/sender-hyvivyza-7dl4s-sender-hyvivyza.log

Checking the broker ingress pod:
https://pantheon.corp.google.com/storage/browser/_details/knative-prow/pr-logs/pull/google_knative-gcp/1415/pull-google-knative-gcp-wi-tests/1281267387093487616/artifacts/controller-logs/cloud-run-events-default-brokercell-ingress-765fb65c68-cmxhw-ingress.txt, it showed error: 
```
{"level":"error","ts":"2020-07-09T17:05:30.178Z","logger":"broker-ingress","caller":"ingress/handler.go:168","msg":"Error publishing to PubSub for broker test-g-c-p-broker-tracing-kqp86/gcp-haoptzdy. event: Validation: valid\nContext Attributes,\n  specversion: 1.0\n  type: e2e-dummy-event-type\n  source: e2e-dummy-event-source\n  id: e2e-dummy-event-id\n  time: 2020-07-09T17:05:30.029102539Z\n  datacontenttype: application/json\nExtensions,\n  knativearrivaltime: 2020-07-09T17:05:30.052197779Z\n  traceparent: 00-a7690e3cd459d2676d06af82e3cc8123-e8526b5409da5713-01\nData,\n  \"{\\\"source\\\": \\\"sender!\\\"}\"\n, err: rpc error: code = Unauthenticated desc = transport: compute: Received 403 `\nUnable to generate token; IAM returned 403 Forbidden: The caller does not have permission\n\nThis error could be caused by a missing IAM policy binding on the target IAM service account.\n\nYou can create the necessary policy binding with:\n\n  gcloud iam service-accounts add-iam-policy-binding \\\n    --role=roles/iam.workloadIdentityUser \\\n    --member=\"serviceAccount:knative-boskos-30.svc.id.goog[cloud-run-events/broker]\" \\\n    prow-job@knative-tests.iam.gserviceaccount.com\n\nFor more information, refer to the Workload Identity documentation:\n\n    https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity\n\n`.","commit":"469cc16","stacktrace":"github.com/google/knative-gcp/pkg/broker/ingress.(*Handler).ServeHTTP\n\tgithub.com/google/knative-gcp/pkg/broker/ingress/handler.go:168\ngo.opencensus.io/plugin/ochttp.(*Handler).ServeHTTP\n\tgo.opencensus.io@v0.22.4/plugin/ochttp/server.go:92\nnet/http.serverHandler.ServeHTTP\n\tnet/http/server.go:2807\nnet/http.(*conn).serve\n\tnet/http/server.go:1895"}
```

As it is related on the flakiness, I include the 500 error here.

Also, in that Ingress Pod log, StorageSource and PubsubSource seem to have the same credential issue, but looks like the source itself will automatically retry, so they finally succeeded.

Reply:
Thanks for adding a TODO, we should figure out how this case should be handled so that the broker does not return 500.